exporters/etw/include/opentelemetry/exporters/etw/etw_provider_exporter.h

```diff
@@ -0,0 +1,657 @@
+        {
+          auto temp  = nostd::get<uint32_t>(value);
+          jObj[name] = temp;
+          break;
+        }
+        case common::AttributeType::TYPE_UINT64:
+        {
+          auto temp  = nostd::get<uint64_t>(value);
+          jObj[name] = temp;
+          break;
+        }
+        case common::AttributeType::TYPE_DOUBLE:
+        {
+          auto temp  = nostd::get<double>(value);
+          jObj[name] = temp;
+          break;
+        }
+        case common::AttributeType::TYPE_STRING:
+        {
+          auto temp  = nostd::get<nostd::string_view>(value);
+          jObj[name] = temp;
+          break;
+        }
+        case common::AttributeType::TYPE_CSTRING:
+        {
+          auto temp  = nostd::get<const char *>(value);
+          jObj[name] = temp;
+          break;
+        }
+#  if HAVE_TYPE_GUID
+          // TODO: consider adding UUID/GUID to spec
+        case common::AttributeType::TYPE_GUID:
+        {
+          auto temp = nostd::get<GUID>(value);
+          // TODO: add transform from GUID type to string?
+          jObj[name] = temp;
+          break;
+        }
+#  endif
+        // TODO: arrays are not supported yet
+#  if 0
+        // TODO: array of uint8_t is not supported by OT spec
+        case common::AttributeType::TYPE_SPAN_BYTE:
+#  endif
+        case common::AttributeType::TYPE_SPAN_BOOL:
+        case common::AttributeType::TYPE_SPAN_INT:
+        case common::AttributeType::TYPE_SPAN_INT64:
+        case common::AttributeType::TYPE_SPAN_UINT:
+        case common::AttributeType::TYPE_SPAN_UINT64:
+        case common::AttributeType::TYPE_SPAN_DOUBLE:
+        case common::AttributeType::TYPE_SPAN_STRING:
+        default:
+          // TODO: unsupported type
+          break;
+      }
+    };
+
+    // Layer 1
+    nlohmann::json l1 = nlohmann::json::array();
+    // Layer 2
+    nlohmann::json l2 = nlohmann::json::array();
+    // Layer 3
+    nlohmann::json l3 = nlohmann::json::array();
+
+    l1.push_back("Span");
+
+    {
+      // TODO: clarify why this is needed
+      // TODO: fix time here
+      nlohmann::json j;
+      j["TypeCode"] = 255;
+      j["Body"]     = "0xFFFFFC60000000005F752C2C";
+      l3.push_back(j);
+    };
+
+    // Actual value object goes here
+    l3.push_back(jObj);
+
+    l2.push_back(l3);
+    l1.push_back(l2);
+
+    {
+      // Another time field again, but at the top
+      // TODO: fix time here
+      nlohmann::json j;
+      j["TypeCode"] = 255;
+      j["Body"]     = "0xFFFFFC60000000005F752C2C";
+      l1.push_back(j);
+    };
+
+    std::vector<uint8_t> v = nlohmann::json::to_msgpack(l1);
+
+    // NUL-terminator and padding for odd-sized buffers
+    v.push_back(0);
+    v.push_back(0);
+    if (v.size() % 2)
+    {
+      v.push_back(0);
+    };
+    void *buff = v.data();
+
+    UCHAR level        = 0;  // LogAlways
+    auto writeResponse = EventWriteString(providerData.providerHandle, level, 0, (PCWSTR)buff);
```

Comment:
Are you sure we can use `EventWriteString` here? MsgPack encoded data might well contain sequences of zeros, which might be interpreted as end of the string. Maybe `EventWrite` is a safer choice here?

Reply:
It's a good question and a valid concern. We may hit this with a sequence of two zeroes: since the originating string is wide string.. I'd like to merge this, but keep that question of 0 0 as an open issue - to see what are the use-cases where MsgPack would result in a payload with two consecutive zeroes. As an alternative, the code may also use Trace Logging Dynamic logic (that would be safe). But we have not finished the open sourcing of TLD header.

Reply:
..basically in case if the payload ended up having two consecutive zeroes, it appears that resulting event may get truncated by receiver. I need to see if we can reuse `EventWrite` (and how). I'll handle it in a separate PR.

Reply:
Thanks!

It's not too hard to come up with scenarios with two consecutive zero bytes (the [msgpack site](https://msgpack.org) has a nice converter to verify those). `65536` encodes to `ce 00 01 00 00`, `[0,0]` encodes to `92 00 00`, and `40000000000` encodes to `cf 00 00 00 09 50 2f 90 00`.

Reply:
I logged a separate issue for this. #483 - side-effect is that the receiving end in this case will get a smaller buffer and would fail to decode it. `EventWrite` is the way to go. But it'd need a bit extra prep-work - to generate the proper struct that describes the event. It'd need about 20-30 lines of code patch. I'm afraid this is too big ask for Mishal, and I'll cover it in a subsequent PR.